### PR TITLE
Refactor StorageCommunicationEndpointProvider to return StorageCommunicationEndpoint per StoreInfo

### PR DESCRIPTION
**Pull request by mmandlis**
This patch moves the StoreInfo argument from the
getStorageEndpointProvider() call to the getStorageEndpoint() call, which means that
a single StorageEndpointProvider can be used for a set of StorageProxies.

Eventually this single StorageEndpointProvider will become a separate
StorageEndpointManager, rather than being a responsibility of Stores directly.
This will allow the same manager to be used in both direct communication mode
and in hosted (via PEC) mode.

A short-term consequence is that both StoreInfo and an ActiveStore are passed
into the StorageProxy constructor (the ActiveStore is currently still the
StorageEndpointProvider and StoreInfo is needed to get the endpoint). This
duplication will be resolved in a subsequent patch (the ActiveStore argument
will become a StorageEndpoint instead, vended by a StorageEndpointManager
which will have responsibility for creating ActiveStores on demand).
________________________________________________________________________________________________

next steps:
- DirectStorageEndpointManager should implement StorageCommunicationEndpointProvider, and ActiveStore StorageCommunicationEndpoint instead.

### CODE REVIEW

File: src/planning/plan/planificator.ts [hunk 1]
```diff
@@ -192,7 +192,7 @@ export class Planificator {
   }
 
   async _storeSearch(): Promise<void> {
-    const handle = handleForActiveStore(this.searchStore, this.arc);
+    const handle = handleForActiveStore(this.searchStore.storeInfo, this.searchStore, this.arc);
```
Comment (shans) [inline src/planning/plan/planificator.ts:195]: Is there any way to avoid having to pass both the store and store.storeInfo? Seems like a recipe for problems later.

**Review by shans — COMMENTED** [review]
Is StoreInfo plumbed all the way through here just for type resolution purposes? If so, I don't think we should do that.

If you do need the data somewhere, could you point out where? It would  be good to capture this kind of information about the change in the change description too, so others can access this information when looking through change history.

**Comment by mmandlis** [comment]
> Is StoreInfo plumbed all the way through here just for type resolution purposes? If so, I don't think we should do that.
> 
> If you do need the data somewhere, could you point out where? It would be good to capture this kind of information about the change in the change description too, so others can access this information when looking through change history.

Yes, my plan was to make the DirectStorageManager an implementation of StorageEndpointCommunicationProvider (instead of ActiveStore), and then the StorageEndpointCommunicationProvider itself doesn't need to be templated, only its getStorageEndpoint method. And then i need the Type to determine the return CRDTRecord's type.

File: src/planning/plan/planificator.ts (hunk 1 above)
Comment (mmandlis) [reply src/planning/plan/planificator.ts:195]: this is an incremental step towards not passing the store itself, but the DirectStoreManager instead (and the storeInfo)

**Comment by shans** [comment]
That sounds like using an argument purely for type resolution, is this correct? i.e. the value isn't used anywhere in the computation, but the type is used by the compiler?

If so I think that's somewhat of an antipattern.

**Comment by mmandlis** [comment]
> That sounds like using an argument purely for type resolution, is this correct? i.e. the value isn't used anywhere in the computation, but the type is used by the compiler?
> 
> If so I think that's somewhat of an antipattern.

There is a code path where the type will actually be used - that is if the ActiveStore needs to be constructed by the StorageManager. 
Either way, what would be a good pattern here?

**Comment by shans** [comment]
I need to see a design doc (and I think this isn't the first time I've asked for one).

Include:
* Where did we start (with diagrams), where are we going (with diagrams), what does the Kotlin stack look like (with diagrams, for reference).

Questions I started asking myself while trying to puzzle through this CL:
 - what is the StorageManager? Why does it constructing ActiveStores have anything to do with needing to pass StoreInfo into the StorageProxy constructor?
 - if it's ActiveStore / DirectStorageManager that are (will be) implementations of StorageEndpoingCommunicationProvider, why did the diagram you sent the other day have StorageCommunicationEndpointProvider attached to the StorageProxy?

In short: there needs to be a clear, consistent and complete description of what you're trying to do before I can have a hope of deciding whether these CLs are going in the right direction.

**Review by shans — APPROVED** [review]
